src/Control/Carrier/Telemetry.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Control.Carrier.Telemetry (
```

Comment:
I'm gonna need extra time to review this file.  We use a lot of threading utils in here, and I really want to make sure we get those right.